### DISCUSSION TITLE
How to use the Github API to get metadata on a repo

### DISCUSSION THREAD

**Discussion by jhan-sims** [community/community#49744] [Apps, API and Webhooks]
### Select Topic Area

Question

### Body

Hi,

Not sure if this is too simple a question but I am quite new to API scripting.

I am trying to get metadata on a private repo using Github's API. I've tried three methods:
1. Windows CMD. `curl -H "Authorization: token [MY_TOKEN]" https://api.github.com/repos/OWNER/REPO` but this outputs "message": "Not Found"

2. Python script.
```
import requests

headers = {'Authorization': 'token [MY_TOKEN]'}

url = f'https://api.github.com/repos/OWNER/REPO'

response = requests.get(url, headers=headers)

if response.status_code == 200:
    repo_data = response.json()
    print(f"Repository name: {repo_data['name']}")
    print(f"Repository description: {repo_data['description']}")
else:
    print(f"Error: {response.status_code} - {response.text}")
```
but this only says Error: 404 - {"message":"Not Found" as well.

3. Lastly, following Github's documentation https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content
```
curl -L \
> -H "Accept: application/vnd.github+json" \
> -H "Authorization: Bearer [MY_TOKEN]"\                                       
> -H "X-GitHub-Api-Version: 2022-11-28" \
> https://api.github.com/repos/OWNER/REPO/contents
```
which outputs curl: (3) URL using bad/illegal format or missing URL
{
  "message": "Bad credentials"
  
I'm sure I'm missing something simple, but if anyone knows how to use Github's API to retrieve metadata on a private repo please let me know.

**Comment by jge162** [comment]
maybe this is issue? 
`> -H "Authorization: Bearer [MY_TOKEN]"\`

where is the space? for the backslash.

**Reply by jhan-sims** [reply]
oo that definitely helped, but it only changed the response from "message": "Bad credentials" to "message": "Not Found"

thank you for that though

**Comment by itspasindu** [comment]
Here are some things to consider:

    Make sure you have the correct permissions: To access a private repository using the GitHub API, you need to have the appropriate permissions. Check that the personal access token you are using has the required permissions to read metadata from the repository.

    Check the repository URL: Make sure that the URL you are using to access the repository is correct. Double-check that the repository name and owner are spelled correctly and that the repository exists.

    Use the correct endpoint: The endpoint you are using in your requests may not be correct. For example, to retrieve metadata on a repository, you should use the /repos/{owner}/{repo} endpoint, where {owner} is the repository owner's GitHub username or organization name, and {repo} is the repository's name.

    Check the API version: GitHub's API is versioned, so you need to specify which version you are using in your request headers. Make sure that the API version you are using matches the version specified in the documentation.

    Check the syntax of your commands: Make sure that you are using the correct syntax when making API requests. For example, in your third approach, there seems to be a syntax error in the command, with the \ character being placed before a space. Try removing the \ character and running the command again.

Here is an example Python script that should retrieve metadata on a private repository using the /repos/{owner}/{repo} endpoint:

python

import requests

```
url = 'https://api.github.com/repos/{owner}/{repo}'
headers = {'Authorization': 'token [MY_TOKEN]', 'Accept': 'application/vnd.github+json'}
params = {'owner': 'OWNER', 'repo': 'REPO'}

response = requests.get(url.format(**params), headers=headers)

if response.status_code == 200:
    repo_data = response.json()
    print(f"Repository name: {repo_data['name']}")
    print(f"Repository description: {repo_data['description']}")
else:
    print(f"Error: {response.status_code} - {response.text}")
```

Make sure to replace `[MY_TOKEN]`, OWNER, and REPO with the appropriate values for your repository and access token.

I hope this helps! Let me know if you have any further questions.

**Comment by jhan-sims** [comment]
for anyone curious, the Python script I posted works, as well as Pasindu's. Just make sure to use a classic token instead of the fine-grained

**Comment by Aidajafarbigloo** [comment]
@jhan-sims Hello

I have the same problem, but even when I use a classic token I get error 404.

**Reply by caviri** [reply]
Hello @Aidajafarbigloo, same issue here. Did you find a solution?

**Reply by Aidajafarbigloo** [reply]
Hello @caviri, Not yet, the same code doesn't work to get metadata for private repos.

**Comment by tobi4120** [comment]
I had the same issue, I was using a user access token from my GitHub app. The issue was that the account I was trying to get the repo from did not have my GitHub app installed, only authorized. Authorization only allows the app to access 'Account permissions'. You need to install the app to grant 'Repository permissions'. After installing, GitHub asks you to select which repositories you give the app permission to access.
